### DISCUSSION TITLE
feature request: uri library

### DISCUSSION THREAD

**Discussion by matteoredaelli** [trealla-prolog/trealla-prolog#689] [General]
Hello

Most of the modern programming languages have a uri/url library for managing urls

Also swi prolog has it: an old one written in prolog and a new one written in c

https://www.swi-prolog.org/pldoc/man?section=url
https://www.swi-prolog.org/pldoc/man?section=uri

Could you add it to Tralla prolog, please?

Thanks in advance
Matteo

**Comment by infradig** [comment]
Good idea.

**Comment by infradig** [comment]
Implementing parse_url/2 would be a start.

**Comment by infradig** [comment]
Added...

	$ tpl
	?- parse_url('http://www.xyz.org:81/hello?msg=Hello+World%21&foo=bar#xyz',P).
	   P = [search([msg='Hello World!',foo=bar]),protocol(http),host('www.xyz.org'),port(81),path('/hello'),fragment(xyz)].

	?- parse_url(U,[search([msg='Hello World!',foo=bar]),protocol(http),host('www.xyz.org'),port(81),path('/hello'),fragment(xyz)]).
	   U = 'http://www.xyz.org:81/hello?msg=Hello+World%21&foo=bar#xyz'.
	?-

I'll see about the rest.

**Comment by triska** [comment]
Wouldn't strings be a lot more appropriate for such ephemeral data, like `"Hello World!"`? There is little chance that other parts of the program need this string (in the sense that it is unlikely to occur anywhere else), and strings can be reclaimed a lot more efficiently on backtracking.

**Comment by infradig** [comment]
Probably, though they are not created as interned atoms, hence just as
easily reclaimed.

**Comment by infradig** [comment]
Added library(uri) as per SWI Prolog.
